server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -415,6 +426,10 @@ static void parseObjectOrField(DocumentParserContext context, Mapper mapper) thr
         }
     }
 
+    private static boolean canToBeFlatten(DocumentParserContext context) {
```

Comment:
I decided to extract this method to make the if-elif-else just above much easier to read.